File: flasher/downloader.py
"""
downloader.py - OTA firmware downloader for OnePlus / OPPO devices.

Resolves OTA download links (including 4PDA redirects) to direct CDN URLs
and downloads them via aria2c for maximum speed.

Requires: aria2c  (sudo pacman -S aria2 / sudo apt install aria2)
"""

import urllib.parse
import logging
from pathlib import Path
from dataclasses import dataclass

from flasher.utils import run_cmd, require_tools

log = logging.getLogger(__name__)

_OTA_HEADERS = [
    "userId: oplus-ota|16002018",
    "User-Agent: okhttp/4.9.2",
    "Accept: application/json",
]


# ---------------------------------------------------------------------------
# Result
# ---------------------------------------------------------------------------

@dataclass
class DownloadResult:
    success: bool
    url: str = ""
    cdn_url: str = ""
    output_path: Path | None = None
    error: str = ""


# ---------------------------------------------------------------------------
# Link resolution
# ---------------------------------------------------------------------------

def _extract_real_url(url: str) -> str:
    """Unwrap 4PDA redirect to get the real OTA endpoint."""
    parsed = urllib.parse.urlparse(url)
    if "4pda.to" in parsed.netloc:
        qs = urllib.parse.parse_qs(parsed.query)
        if "u" in qs:
            real = urllib.parse.unquote(qs["u"][0])
            log.info(f"Unwrapped 4PDA redirect -> {real}")
            return real
    return url


def _resolve_cdn(url: str) -> str | None:
    """
    Follow the OTA downloadCheck endpoint to get the real CDN URL.

    The server returns HTTP 302 with the CDN link in Location header.
    Uses curl to avoid pulling in requests as a dependency.
    """
    # Build curl command with custom headers, follow no redirects, dump headers only
    cmd = ["curl", "-s", "-o", "/dev/null", "-D", "-", "--max-redirs", "0"]
    for h in _OTA_HEADERS:
        cmd += ["-H", h]
    cmd.append(url)

    rc, stdout, stderr = run_cmd(cmd, timeout=30)

    if rc not in (0, 6, 7):  # curl exits non-zero on redirect when --max-redirs 0
        # Check if we got a Location header anyway (some curl versions)
        pass

    for line in stdout.splitlines():
        if line.lower().startswith("location:"):
            cdn = line.split(":", 1)[1].strip()
            log.info(f"CDN URL resolved: {cdn}")
            return cdn

    log.error(f"Could not find Location header in response:\n{stdout[:500]}")
    return None


# ---------------------------------------------------------------------------
# Download
# ---------------------------------------------------------------------------

def download_firmware(
    url: str,
    output_dir: Path | None = None,
    connections: int = 16,
) -> DownloadResult:
    """
    Resolve an OTA link and download the firmware via aria2c.

    Args:
        url:         OTA link (direct or 4PDA redirect).
        output_dir:  Directory to save the file (default: cwd).
        connections: Number of parallel connections for aria2c.

    Returns:
        DownloadResult dataclass.
    """
    if not require_tools("aria2c", "curl"):
        return DownloadResult(success=False, url=url,
                              error="Missing required tools: aria2c, curl")

    # Step 1: unwrap 4PDA redirect if needed
    real_url = _extract_real_url(url)

    print(f"\n  OTA endpoint : {real_url}")

    # Step 2: resolve to CDN
    print("  Resolving CDN link ...")
    cdn_url = _resolve_cdn(real_url)
    if not cdn_url:
        return DownloadResult(success=False, url=real_url,
                              error="Failed to resolve CDN URL (no Location header in response)")

    print(f"  CDN URL      : {cdn_url}")

    # Show link expiry if present in query string (e.g. ?X-Amz-Expires=... or ?e=...)
    try:
        import urllib.parse, datetime
        qs = urllib.parse.parse_qs(urllib.parse.urlparse(cdn_url).query)
        # AWS-style: X-Amz-Date + X-Amz-Expires
        if "X-Amz-Date" in qs and "X-Amz-Expires" in qs:
            amz_date = qs["X-Amz-Date"][0]  # e.g. 20240101T120000Z
            expires_s = int(qs["X-Amz-Expires"][0])
            created = datetime.datetime.strptime(amz_date, "%Y%m%dT%H%M%SZ").replace(
                tzinfo=datetime.timezone.utc
            )
            expires_at = created + datetime.timedelta(seconds=expires_s)
            now = datetime.datetime.now(datetime.timezone.utc)
            remaining = expires_at - now
            if remaining.total_seconds() > 0:
                h, m = divmod(int(remaining.total_seconds()) // 60, 60)
                print(f"  Link expires : {expires_at.strftime('%Y-%m-%d %H:%M UTC')}  ({h}h {m}m remaining)")
            else:
                print(f"  Link expires : EXPIRED ({expires_at.strftime('%Y-%m-%d %H:%M UTC')})")
        # Generic Unix timestamp: ?e=1234567890
        elif "e" in qs:
            ts = int(qs["e"][0])
            expires_at = datetime.datetime.fromtimestamp(ts, tz=datetime.timezone.utc)
            now = datetime.datetime.now(datetime.timezone.utc)
            remaining = expires_at - now
            if remaining.total_seconds() > 0:
                h, m = divmod(int(remaining.total_seconds()) // 60, 60)
                print(f"  Link expires : {expires_at.strftime('%Y-%m-%d %H:%M UTC')}  ({h}h {m}m remaining)")
            else:
                print(f"  Link expires : EXPIRED ({expires_at.strftime('%Y-%m-%d %H:%M UTC')})")
    except Exception:
        pass

    # Step 3: download via aria2c
    cmd = [
        "aria2c",
        f"-x{connections}",
        f"-s{connections}",
        "-k", "1M",
        "--file-allocation=none",
        "--console-log-level=notice",
    ]
    if output_dir:
        output_dir = Path(output_dir)
        output_dir.mkdir(parents=True, exist_ok=True)
        cmd += ["-d", str(output_dir)]

    cmd.append(cdn_url)

    print(f"\n  Starting download ({connections} connections) ...\n")

    # Run aria2c interactively so the user sees the progress bar
    import subprocess
    proc = subprocess.run(cmd)

    if proc.returncode != 0:
        return DownloadResult(
            success=False, url=real_url, cdn_url=cdn_url,
            error=f"aria2c exited with code {proc.returncode}",
        )

    # Try to figure out where aria2c saved the file
    filename = cdn_url.split("/")[-1].split("?")[0]
    output_path = (Path(output_dir) / filename) if output_dir else Path.cwd() / filename

    return DownloadResult(
        success=True,
        url=real_url,
        cdn_url=cdn_url,
        output_path=output_path if output_path.exists() else None,
    )

File: src/downloader.rs
//! OTA firmware downloader for OnePlus / OPPO / Realme devices.
//!
//! Resolves OTA download links (including 4PDA redirects) to direct CDN URLs
//! and downloads them via aria2c for maximum speed.

use std::path::{Path, PathBuf};
use std::process::Command;

use crate::utils::{require_tools, run_cmd};
use log::{error, info};

const OTA_HEADERS: &[&str] = &[
    "userId: oplus-ota|16002018",
    "User-Agent: okhttp/4.9.2",
    "Accept: application/json",
];

#[derive(Debug, Clone)]
pub struct DownloadResult {
    pub success: bool,
    pub url: String,
    pub cdn_url: String,
    pub output_path: Option<PathBuf>,
    pub error: String,
}

impl DownloadResult {
    fn fail(url: &str, error: &str) -> Self {
        Self {
            success: false,
            url: url.into(),
            cdn_url: String::new(),
            output_path: None,
            error: error.into(),
        }
    }
}

/// Unwrap 4PDA redirect to get real OTA endpoint.
fn extract_real_url(url: &str) -> String {
    if let Ok(parsed) = url::Url::parse(url) {
        if parsed
            .host_str()
            .map(|h| h.contains("4pda.to"))
            .unwrap_or(false)
        {
            for (k, v) in parsed.query_pairs() {
                if k == "u" {
                    info!("Unwrapped 4PDA redirect -> {}", v);
                    return v.to_string();
                }
            }
        }
    }
    url.to_string()
}

/// Follow OTA server 302 redirect to get CDN URL.
fn resolve_cdn(url: &str) -> Option<String> {
    let mut parts: Vec<String> = vec![
        "curl".into(),
        "-s".into(),
        "-o".into(),
        "/dev/null".into(),
        "-D".into(),
        "-".into(),
        "--max-redirs".into(),
        "0".into(),
    ];
    for h in OTA_HEADERS {
        parts.push("-H".into());
        parts.push(h.to_string());
    }
    parts.push(url.into());
    let refs: Vec<&str> = parts.iter().map(|s| s.as_str()).collect();
    let r = run_cmd(&refs, 30);
    for line in r.stdout.lines() {
        if line.to_lowercase().starts_with("location:") {
            let cdn = line.split_once(':').map(|(_, v)| v.trim().to_string())?;
            info!("CDN URL resolved: {}", cdn);
            return Some(cdn);
        }
    }
    error!("Could not find Location header in response");
    None
}

/// Resolve OTA link and download firmware via aria2c.
pub fn download_firmware(url: &str, output_dir: Option<&Path>, connections: u32) -> DownloadResult {
    if !require_tools(&["aria2c", "curl"]) {
        return DownloadResult::fail(url, "Missing tools: aria2c, curl");
    }

    let real_url = extract_real_url(url);
    println!("\n  OTA endpoint : {}", real_url);
    println!("  Resolving CDN link ...");

    let cdn_url = match resolve_cdn(&real_url) {
        Some(c) => c,
        None => return DownloadResult::fail(&real_url, "Failed to resolve CDN URL"),
    };
    println!("  CDN URL      : {}", cdn_url);

    // Show link expiry if available
    if let Ok(parsed) = url::Url::parse(&cdn_url) {
        let params: std::collections::HashMap<_, _> = parsed.query_pairs().collect();
        if let Some(ts) = params.get("e") {
            if let Ok(ts) = ts.parse::<u64>() {
                let now = std::time::SystemTime::now()
                    .duration_since(std::time::UNIX_EPOCH)
                    .unwrap_or_default()
                    .as_secs();
                if ts > now {
                    println!(
                        "  Link expires : in {}h {}m",
                        (ts - now) / 3600,
                        ((ts - now) % 3600) / 60
                    );
                } else {
                    println!("  Link expires : EXPIRED");
                }
            }
        }
    }

    let mut cmd = Command::new("aria2c");
    cmd.arg(format!("-x{}", connections))
        .arg(format!("-s{}", connections))
        .arg("-k")
        .arg("1M")
        .arg("--file-allocation=none")
        .arg("--console-log-level=notice");
    if let Some(dir) = output_dir {
        std::fs::create_dir_all(dir).ok();
        cmd.arg("-d").arg(dir);
    }
    cmd.arg(&cdn_url);

    println!("\n  Starting download ({} connections) ...\n", connections);
    let status = match cmd.status() {
        Ok(s) => s,
        Err(e) => {
            return DownloadResult {
                success: false,
                url: real_url,
                cdn_url,
                output_path: None,
                error: format!("aria2c: {}", e),
            };
        }
    };
    if !status.success() {
        return DownloadResult {
            success: false,
            url: real_url,
            cdn_url,
            output_path: None,
            error: format!("aria2c exited {}", status.code().unwrap_or(-1)),
        };
    }

    let filename = cdn_url
        .split('/')
        .last()
        .unwrap_or("firmware.zip")
        .split('?')
        .next()
        .unwrap_or("firmware.zip");
    let out = output_dir
        .map(|d| d.join(filename))
        .unwrap_or_else(|| std::env::current_dir().unwrap_or_default().join(filename));
    DownloadResult {
        success: true,
        url: real_url,
        cdn_url,
        output_path: if out.exists() { Some(out) } else { None },
        error: String::new(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_extract_4pda() {
        assert_eq!(
            extract_real_url("https://4pda.to/goto/?u=https%3A%2F%2Fexample.com%2Ffw.zip"),
            "https://example.com/fw.zip"
        );
    }
    #[test]
    fn test_extract_direct() {
        assert_eq!(
            extract_real_url("https://example.com/fw.zip"),
            "https://example.com/fw.zip"
        );
    }
}
